blocks/main.go

```diff
@@ -0,0 +1,430 @@
+	  writer_id STRING NOT NULL,
+	  block_num BIGINT NOT NULL,
+	  raw_bytes BYTES NOT NULL,
+	  PRIMARY KEY (block_id, writer_id, block_num)
+	)`); err != nil {
+		return nil, err
+	}
+
+	if *splits > 0 {
+		r := rand.New(rand.NewSource(int64(time.Now().UnixNano())))
+		for i := 0; i < *splits; i++ {
+			if _, err := db.Exec(`ALTER TABLE blocks SPLIT AT ($1, '', 0)`, r.Int63()); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return &cockroach{db: db}, nil
+}
+
+type mongo struct {
+	blocks *mgo.Collection
+}
+
+func (m *mongo) write(writerID string, blockNum int64, blockCount int, r *rand.Rand) error {
+	type Block struct {
+		BlockID  int64
+		WriterID string
+		BlockNum int64
+		RawBytes []byte
+	}
+	docs := make([]interface{}, blockCount)
+	for i := 0; i < blockCount; i++ {
+		docs[i] = &Block{
+			BlockID:  r.Int63(),
+			WriterID: writerID,
+			BlockNum: blockNum + int64(i),
+			RawBytes: randomBlock(r),
+		}
+	}
+
+	return m.blocks.Insert(docs...)
+}
+
+func setupMongo(parsedURL *url.URL) (database, error) {
+	session, err := mgo.Dial(parsedURL.String())
+	if err != nil {
+		panic(err)
+	}
+
+	session.SetMode(mgo.Monotonic, true)
+	// session.SetSafe(&mgo.Safe{FSync: true})
+
+	blocks := session.DB("datablocks").C("blocks")
+	return &mongo{blocks: blocks}, nil
+}
+
+type cassandra struct {
+	session *gocql.Session
+}
+
+func (c *cassandra) write(writerID string, blockNum int64, blockCount int, r *rand.Rand) error {
+	const insertBlockStmt = `INSERT INTO datablocks.blocks ` +
+		`(block_id, writer_id, block_num, raw_bytes) VALUES (?, ?, ?, ?)`
+	return c.session.Query(insertBlockStmt, r.Int63(), writerID, blockNum, randomBlock(r)).Exec()
+}
+
+func setupCassandra(parsedURL *url.URL) (database, error) {
+	if *batch != 1 {
+		return nil, fmt.Errorf("unsupported --batch size: %d", *batch)
+	}
+
+	cluster := gocql.NewCluster(parsedURL.Host)
+	cluster.Consistency = gocql.Quorum
+	s, err := cluster.CreateSession()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	const createKeyspace = `
+CREATE KEYSPACE IF NOT EXISTS datablocks WITH REPLICATION = {
+  'class' : 'SimpleStrategy',
+  'replication_factor' : 1
+};`
+	const createTable = `
+CREATE TABLE IF NOT EXISTS datablocks.blocks(
+  block_id BIGINT,
+  writer_id TEXT,
+  block_num BIGINT,
+  raw_bytes BLOB,
+  PRIMARY KEY(block_id, writer_id, block_num)
+);`
+
+	if err := s.Query(createKeyspace).RetryPolicy(nil).Exec(); err != nil {
+		log.Fatal(err)
+	}
+	if err := s.Query(createTable).RetryPolicy(nil).Exec(); err != nil {
+		log.Fatal(err)
+	}
+	return &cassandra{session: s}, nil
+}
+
+// setupDatabase performs initial setup for the example, creating a database and
+// with a single table. If the desired table already exists on the cluster, the
+// existing table will be dropped.
+func setupDatabase(dbURL string) (database, error) {
+	parsedURL, err := url.Parse(dbURL)
+	if err != nil {
+		return nil, err
+	}
+	parsedURL.Path = "datablocks"
+
+	switch parsedURL.Scheme {
+	case "postgresql":
```

Comment:
we've been using `postgres` everywhere in the url. It's probably best to handle both here.